src/ILCompiler.Compiler/src/Compiler/CompilerTypeSystemContext.EnumMethods.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+
+using Internal.TypeSystem;
+using Internal.IL.Stubs;
+
+using Debug = System.Diagnostics.Debug;
+
+namespace ILCompiler
+{
+    partial class CompilerTypeSystemContext
+    {
+        private class EnumInfo
+        {
+            public TypeDesc Type => EqualsMethod.OwningType;
+
+            public MethodDesc EqualsMethod { get; }
+
+            public MethodDesc GetHashCodeMethod { get; }
+
+            public EnumInfo(TypeDesc enumType)
+            {
+                Debug.Assert(enumType.IsEnum && enumType.IsTypeDefinition);
+                EqualsMethod = new EnumEqualsThunk(enumType);
+                GetHashCodeMethod = new EnumGetHashCodeThunk(enumType);
+            }
+        }
+
+        private class EnumInfoHashtable : LockFreeReaderHashtable<TypeDesc, EnumInfo>
+        {
+            protected override int GetKeyHashCode(TypeDesc key) => key.GetHashCode();
+            protected override int GetValueHashCode(EnumInfo value) => value.Type.GetHashCode();
+            protected override bool CompareKeyToValue(TypeDesc key, EnumInfo value) => key == value.Type;
+            protected override bool CompareValueToValue(EnumInfo v1, EnumInfo v2) => v1.Type == v2.Type;
+
+            protected override EnumInfo CreateValueFromKey(TypeDesc key)
+            {
+                return new EnumInfo(key);
+            }
+        }
+
+        private EnumInfoHashtable _enumInfoHashtable = new EnumInfoHashtable();
+
+        public MethodDesc TryResolveConstrainedEnumMethod(TypeDesc enumType, MethodDesc virtualMethod)
```

Comment:
This api shouldn't be part of the type system context that is specific to the compiler. We may wish to use it elsewhere. Please make it either an api surface exposed on the basic TypeSystemContext, or just make a separate non-typesystem joined api for this.